src/replication.c

```diff
@@ -218,41 +218,46 @@ perform_read_write_on_fd(int fd, uint8_t *data, uint64_t len, int state)
  * get_all_read_resp_data_chunk will create/update io_data_chunk_list
  * with response received from replica according to io_numer.
  */
-void
-get_all_read_resp_data_chunk(replica_rcomm_resp_t *resp, struct io_data_chunk_list_t *io_chunk_list)
+static void
+get_all_read_resp_data_chunk(replica_rcomm_resp_t *resp, size_t block_len,
+    struct io_data_chunk_list_t *io_chunk_list)
 {
 	zvol_io_hdr_t *hdr = &resp->io_resp_hdr;
 	struct zvol_io_rw_hdr *io_hdr;
 	uint8_t *dataptr = resp->data_ptr;
-	uint64_t len = 0, parsed = 0;
+	uint64_t parsed = 0, data_len;
 	io_data_chunk_t  *io_chunk_entry;
 
 	if (!TAILQ_EMPTY(io_chunk_list)) {
 		io_chunk_entry = TAILQ_FIRST(io_chunk_list);
 
 		while (parsed < hdr->len) {
 			io_hdr = (struct zvol_io_rw_hdr *)dataptr;
-
-			if (io_chunk_entry->io_num < io_hdr->io_num) {
-				io_chunk_entry->data = dataptr + sizeof(struct zvol_io_rw_hdr) + len;
-				io_chunk_entry->io_num = io_hdr->io_num;
-				io_chunk_entry->len = io_hdr->len;
+			data_len = 0;
+			while (data_len < io_hdr->len) {
+				if (io_chunk_entry->io_num < io_hdr->io_num) {
```

Comment:
there can be case where io_num is 0, as replica turned healthy by the time IO reached it, but, target found it to be in degraded when it sent the IO.

Reply:
ok. I have added a flag for reading metadata in zvol_io_hdr. By this we can solve this issue.